evm/vm/base.py

```diff
@@ -86,7 +86,7 @@ def apply_transaction(self, transaction):
         self.block = block
         self.receipts.append(receipt)
 
-        self.clear_journal()
+        self.chaindb.persist()
```

Comment:
If this `persist` is run anywhere between a pair of `id = record()` and `commit(id)` calls, it will break the commit call. What if some caller higher up the stack decides it wants to revert? It just lost the option. `persist()` has the semantics of "commit all in-progress transactions/recordings", which seems problematic anywhere we use it. I don't think we should even have the option to do that.

The more I think about it, the more I don't see a reason to allow committing multiple levels of transactions in one call:
```
record
  record
    record
commit <- I don't think this should be possible. Child transactions are in an unknown state
    commit
  commit
commit
```

Maybe we want to keep `persist` explicit, so we know when we are really writing to the wrapped database. That could make sense. Since the JournalDB implicitly starts with a record, that special outermost transaction can have a special commit: `persist`. So:

```
journal init
  record
    record
    commit
  commit
persist
```

That implies we should make the public `commit()` call fail if there's only one changeset.

Reply:
>`persist()` has the semantics of "commit all in-progress transactions/recordings"

That's true but my understanding was that transaction processing is sequential and self-contained. I didn't see the case of "child transactions" to be honest.

If the processing goes like this, then there shouldn't be a problem I think

```
transaction 1
  commit()
    commit()
    revert()
  commit()
  persist()
transaction 2
  commit()
    commit()
  persist()
transaction 3
 ...
  persist()
```

However, your comment mentioning "child transactions" indicates that my assumption of this sequential processing maybe wrong and that `apply_transaction` may be run stacked (which I actually didn't come across while debugging but I also didn't explicitly pay attention)

Reply:
I think it could be totally reasonable to say that we don't want to support nested transactions. If so, we could simplify several things:

- `JournalDB` writes through all changes until `record()` is called
- once `record()` is called, it cannot be called again until `commit()`
- `commit()` writes through all pending changes
- remove `persist()` as redundant

As I'm typing it out, I realize that's starting to sound a lot like BatchDB. So I'm realizing that I still don't totally get what is the primary goal here and/or what we want different from BatchDB.

Reply:
Judging by your last comment on the main thread you may have already realized what I'm about to write but just for completeness let me elaborate further on this.

The refactored `JournalDB` works very much like the `BatchDB` in the sense that it does not write to the underlying db until some condition is met. This condition could be leaving a `with` statement or calling an explicit API. In the current design, you have to explicitly call `persist()` in order to write to the underlying database.

Before you call `persist` all changes (`set` / `delete`) go into the active changeset. Notice that there is *always* an active changeset where changes end up in.

However, you can build a stack of changesets if you explicitly call `record`. E.g.

```
changeset_1 = db.record()
db.set(a, 1)
changeset_2 = db.record()
db.set(a, 2)
changeset_3 = db.record()
db.delete(b)
db.set(c, 1)
```
At this point you'll end up with a stack of changesets that looks like this:

```
root_changeset
changeset_1
  sets a to 1
changeset_2
  sets a to 2
changset_3
  deletes b
  sets c to 1
```

Calling `commit()` or `discard(changeset_id)` will simply modify how this stack looks like but again, won't do any changes to the underlying db.

For instance, we could call `discard(changeset_3)` and would end up with this:

```
root_changeset
changeset_1
  sets a to 1
changeset_2
  sets a to 2
```

Or we could call `commit(changeset_3)` and would end up with this

```
root_changeset
changeset_1
  sets a to 1
changeset_2
  sets a to 2
  deletes b
  sets c to 1
```

Notice that `commit()` and `discard(changeset_id)` are called as part of the `Computation` handling. E.g. if a computation fails for some reason we discard the changes.

The `persist()` part on the other hand is called on the very end of the transaction processing (when all the computations are done) that's why there is only a single call to `persist()` and I *think* this should be sound.

Reply:
I *think* I'm in support of having this separate mechanism.  We can keep it for now and sort it out in #507 if it proves problematic.